common/kernel/timing.cc

```diff
@@ -616,20 +626,28 @@ void TimingAnalyser::walk_backward()
         auto &dom = domains.at(dom_id);
         for (auto &ep : dom.endpoints) {
             auto &pd = ports.at(ep.first);
-            DelayPair init_setuphold(0);
+            DelayPair init_required(0);
             CellPortKey clock_key;
             // TODO: clock routing delay, if analysis of that is enabled
             if (ep.second != IdString()) {
                 // Add setup/hold time, if this endpoint is clocked
                 for (auto &fanin : pd.cell_arcs) {
-                    if (fanin.type == CellArc::SETUP && fanin.other_port == ep.second)
-                        init_setuphold.min_delay -= fanin.value.maxDelay();
+
+                    if (fanin.type == CellArc::SETUP && fanin.other_port == ep.second) {
+                        if (with_clock_skew) {
+                            auto clock_delay = ports.at(CellPortKey(ep.first.cell, fanin.other_port)).route_delay;
+                            clock_delay.min_delay = clock_delay_fac * clock_delay.min_delay;
+                            clock_delay.max_delay = clock_delay_fac * clock_delay.max_delay;
+                            init_required += clock_delay;
+                        }
+                        init_required.min_delay -= fanin.value.maxDelay();
+                    }
                     if (fanin.type == CellArc::HOLD && fanin.other_port == ep.second)
-                        init_setuphold.max_delay -= fanin.value.maxDelay();
+                        init_required.max_delay += fanin.value.maxDelay();
```

Comment:
I think the hold time had the wrong sign.

Hold slack is calculated like `arr.value.minDelay() - req.value.maxDelay() + clock_to_clock`. If you say here that hold is negative then it becomes `arr.value.minDelay() + hold_time - <rest of required time> + clock_to_clock` which is not correct.

Reply:
Not directly relevant here but some related food for thought - if you use minimum clock/routing delays for the arrival time and maximum for the required time, this is likely to be unnecessarily pessimistic, particularly on the clock side where it is probably that some of the routing is shared so it can't physically have two different delays.

But in general, nextpnr doesn't really have a clear concept of whether min/max delays are overall or for a specific corner (iirc Vivado has fast-min, fast-max, slow-min, slow-max for example which might actually be a better use of the four values DelayQuad than seperate rising/falling delays). This means that as well as shared routing that's guaranteed to have the same delay; as the entire chip will be at essentially the same temperature, voltage and process corner; the difference between delays will never be as high as min vs max.

To avoid this pessimism I think it might be better to just use one corner of routing delays (e.g. min delays everywhere) for hold analysis. Even if this is slightly optimistic it's unlikely for any of the FPGAs we deal with that on-chip variation is going to be that high (but OCV is a field in itself...)

Reply:
Currently it has no impact since the route delay is just a `delay_t` so the min/max corners are the same. But a comment there would be nice if this ever changes.